packages/datadog-instrumentations/src/mocha.js

```diff
@@ -129,6 +160,41 @@ addHook({
   file: 'lib/runner.js'
 }, mochaHook)
 
+addHook({
+  name: 'mocha',
+  versions: ['>=5.2.0'],
+  file: 'lib/runnable.js'
+}, (Runnable) => {
+  shimmer.wrap(Runnable.prototype, 'run', run => function () {
+    const isBeforeEach = this.parent._beforeEach.includes(this)
+    const isAfterEach = this.parent._afterEach.includes(this)
+
+    const isTestHook = isBeforeEach || isAfterEach
+
+    // we restore the original user defined function
+    if (this.fn.asyncResource) {
+      const originalFn = originalFns.get(this.fn)
+      this.fn = originalFn
+    }
+
+    if (isTestHook || this.type === 'test') {
+      const test = isTestHook ? this.ctx.currentTest : this
+      const asyncResource = getTestAsyncResource(test)
+
+      // we bind the test fn to the correct async resource
+      const newFn = asyncResource.bind(this.fn)
+
+      // we store the original function, not to lose it
+      originalFns.set(newFn, this.fn)
+
+      this.fn = newFn
```

Comment:
this piece is so that the test function and hook functions are called with the correct context (we bind the `asyncResource`)